### PR TITLE
perf: replacing network writer with a faster version that can do bit packing

### PR DESCRIPTION
**Pull request by James-Frowen**
BREAKING: NetworkWriter and NetworkReader have been completely re-written to support bitpacking

### CODE REVIEW

**Comment by vladqwerty435** [comment]
**OOOH**

File: Assets/Mirage/Runtime/NetworkPlayer.cs [hunk 1]
```diff
@@ -213,30 +203,36 @@ public void ClearHandlers()
         /// <returns></returns>
         public void Send<T>(T message, int channelId = Channel.Reliable)
         {
+            if (isDisconnected) { return; }
```
Comment (Lymdun) [inline Assets/Mirage/Runtime/NetworkPlayer.cs:206]: Shouldn't such changes be in another PR?

File: Assets/Mirage/Runtime/NetworkPlayer.cs [hunk 2]
```diff
@@ -140,17 +140,7 @@ void AdapterFunction(INetworkPlayer player, NetworkReader reader)
                 // further attacks.
                 var message = default(T);
 
-                // record start position for NetworkDiagnostics because reader might contain multiple messages if using batching
-                int startPos = reader.Position;
```
Comment (James-Frowen) [inline Assets/Mirage/Runtime/NetworkPlayer.cs:144]: this doesn't need to be removed, I will re-add this

**Review by Lymdun — APPROVED** [review]
